mm2src/lp_ordermatch.rs

```diff
@@ -1938,19 +1929,29 @@ impl<Key, Value> TrieDiffHistory<Key, Value> {
 
 type TrieOrderHistory = TrieDiffHistory<Uuid, OrderbookItem>;
 
-#[derive(Default)]
 struct OrderbookPubkeyState {
     /// Timestamp of the latest keep alive message received
     last_keep_alive: u64,
     /// The map storing historical data about specific pair subtrie changes
     /// Used to get diffs of orders of pair between specific root hashes
-    order_pairs_trie_state_history: HashMap<AlbOrderedOrderbookPair, TrieOrderHistory>,
+    order_pairs_trie_state_history: TimeCache<AlbOrderedOrderbookPair, TrieOrderHistory>,
```

Comment:
I remember that we discussed changing the internal `TrieDiffHistory` hash map to the `TimeCache` too. I even think that my initial proposal was wrong, and it might be enough to have `TimeCache` in  TrieDiffHistory only. What do you think?

Reply:
If `TimeCache` is used in `TrieDiffHistory` then the `from_history` function https://github.com/KomodoPlatform/atomicDEX-API/blob/699af01fcc2a3ff3a56f7d0954ef4ba3fda7c750/mm2src/lp_ordermatch.rs#L600 will get orders from history instead of the full trie, which leads to not sending some orders (the orders removed due to `TimeCache` expiration) in the response of `process_sync_pubkey_orderbook_state` function https://github.com/KomodoPlatform/atomicDEX-API/blob/699af01fcc2a3ff3a56f7d0954ef4ba3fda7c750/mm2src/lp_ordermatch.rs#L637 but using `TimeCache` only for `order_pairs_trie_state_history` https://github.com/KomodoPlatform/atomicDEX-API/blob/699af01fcc2a3ff3a56f7d0954ef4ba3fda7c750/mm2src/lp_ordermatch.rs#L1937 will get the orders from the full trie if the pair history expired https://github.com/KomodoPlatform/atomicDEX-API/blob/699af01fcc2a3ff3a56f7d0954ef4ba3fda7c750/mm2src/lp_ordermatch.rs#L657-L659 which should cause no problems

Reply:
> will get orders from history instead of the full trie, which leads to not sending some orders (the orders removed due to TimeCache expiration) in the response of process_sync_pubkey_orderbook_state function 

I've written a test demonstrating that there will be no problem if some records are removed from TrieDiffHistory when expired. https://github.com/KomodoPlatform/atomicDEX-API/compare/memory-leak-fix...KomodoPlatform:memory-leak-fix-triediffhistory-time-cache?expand=1#diff-43046b031e4852b1b604c3583c62cb938c7b04343b859a14fdb286678c3a79ebR2560 (Please merge dev to your branch so it will be more clear what changes I made) 

If Alice state is too outdated, Bob node will remove a certain root hash from history, and then return full trie. Alice will be fully synced - no orders will be lost because full trie returns the *full actual state*. You can imagine this like half of git history is deleted so you have to sync the full repo from scratch.

I've also realized that TrieDiffHistory ttl should be pretty low. The history of very active nodes (like cipi's) can become big quite fast and outweigh the full trie size.